### ISSUE TITLE
tensorrt 7.2.3.4 get some wrong boxes with high score,but tensort 8.2 is correct

### ISSUE THREAD

**Issue by we0091234** [ultralytics/yolov5#6012]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

tensorrt 7.2.3.4 get some wrong boxes with high score,but tensort 8.2 is correct
ubuntu 16.04  tensorrt 7.2.3.4  cuda 10.2  

![zidane](https://user-images.githubusercontent.com/29919617/146485594-c859bde3-2686-4d8a-b8b8-0b9dbc70c817.jpg)

have you tried the  tensorrt 7 ？

**Comment by glenn-jocher** [comment]
@we0091234 thanks for raising this issue! Can you implement a version check with a warning that TensorRT produces incorrect boxes?

**Comment by glenn-jocher** [comment]
@we0091234 I think this is the place:
https://github.com/ultralytics/yolov5/blob/c1249a47c7fe19e2067cb25ed8347e67d26ff1f1/models/common.py#L330

**Comment by glenn-jocher** [comment]
@we0091234 good news 😃! Your original issue may now be fixed ✅ in PR #6021. This PR does not address any TRT 7 issues, but it creates a version check with a warning that TRT 8 is required by YOLOv5.

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by glenn-jocher** [comment]
@we0091234 also I see different ONNX opsets are used in export.py depending on TRT version. Could ONNX Opset 12 be causing the TRT 7 issues?

https://github.com/ultralytics/yolov5/blob/407a9057478d3deea0a9984af42162d21afa2bd2/export.py#L268-L273

**Comment by 1chimaruGin** [comment]
I tried with opset 12 and TRT 7.2.1.3, get some wrong boxes too
![zidane](https://user-images.githubusercontent.com/35499653/147819402-854af18e-b073-49f9-91a3-de942a1a20fd.jpg)
.

**Comment by glenn-jocher** [comment]
@1chimaruGin what happens if you try opset 11 or 13 with TRT 7?

**Comment by 1chimaruGin** [comment]
@glenn-jocher 
With opset 11, also has wrong boxes too. 
With opset 13, can't export TensorRT.

**Comment by glenn-jocher** [comment]
@1chimaruGin ok, then it seems like we need to set a lower limit on TRT>=8.0 with YOLOv5. I'll consult the TRT export author and add a TODO to investigate

**Comment by glenn-jocher** [comment]
@imyhxy it seems like TRT 7 export causes problems (see above). 

Do you think we should implement a hard requirement for TRT 8.0 export and inference with YOLOv5? I implemented a warning (but not an error) on TRT 7 inference in https://github.com/ultralytics/yolov5/pull/6021, but this is only a warning on inference, and does not apply to export.

**Comment by imyhxy** [comment]
@glenn-jocher Hi, jocher

I found out that there are some weird issues between TRT 7 and `anchor_grid` variable, when the `anchor_grid` is expanded to the feature size, the `engine` will output the wrong results (looks like the memory order of `anchor_grid` in TRT 7 is messed). I make a workround patch to the `export_engine` function and will not actually affect the model, check it out in #6185

https://github.com/ultralytics/yolov5/blob/fb839298a16fca4143dc70619a084f9ed085ac07/models/yolo.py#L80

**Comment by we0091234** [comment]
@imyhxy great job!  It's ok now with  tensorrt7.2.3.4

**Comment by glenn-jocher** [comment]
@we0091234 good news 😃! Your original issue may now be fixed ✅ in PR #6185 by @imyhxy. To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by we0091234** [comment]
@glenn-jocher yes ,It works ！

**Comment by glenn-jocher** [comment]
@we0091234 great thanks! TODO removed.

**Comment by 1chimaruGin** [comment]
@glenn-jocher  

While exporting TRT, this error occured.
```
TensorRT: export failure: version 8.0.0 required by YOLOv5, but version 7.1.3.0 is currently installed
```
I think this [line](https://github.com/ultralytics/yolov5/blob/33a67b4918aa3f2e572f115781e615a91fd543a4/export.py#L177) `if trt.__version__[0] == "7"` need to fix.

**Comment by glenn-jocher** [comment]
@1chimaruGin we've been updating this section a lot recently. Can you git pull to make sure your code is up to date and retry your command?

**Comment by glenn-jocher** [comment]
@1chimaruGin good news 😃! Your original issue may now be fixed ✅ in PR #6235. To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
